Detail or remove overflow handling (I needed to know max value of uint to understand commentary.)

waded opened the pull request:
My first reaction when reading "ensures that there are no overflow errors in the future" was "yeah right", because it only checked input.  What I needed to understand is that the amount being minted is limited to a very small amount relative to the maximum value that can be held by `uint` (I did not know what that was off the top of my head) - no overflow is not entirely prevented, but it will not overflow in the first call to mint(), or any reasonable feasible of calls, and so not easily by calls to send() either where there's no such checks. If that's the intent/design, perhaps being more specific as in this edit will help people like me see that right off.

Update: Second option is to remove the LOC purported to handle overflow, simplifying the sample, removing the bump I hit altogether.

waded committed:
Detail the overflow handling here

My first reaction when reading "ensures that there are no overflow errors in the future" was "yeah right", because it only checked input.  What I needed to understand is that the amount being minted is limited to a very small amount relative to the maximum value that can be held by `unit` (I did not know what that was off the top of my head) - no overflow is not entirely prevented, forever, but not in the very first call to mint(), or any reasonable feasible of calls, and so not easily by calls to send() either where there's no such checks. If that's the intent/design, perhaps being more specific as in this edit will help me see that right off.

 If that's correct, please take this improvement.

waded committed:
unit maximum value exceeds 1e77 not 1e71

waded committed:
Fix chk_coding_style

Fix chk_coding_style error "Trailing whitespace found"

chriseth commented on docs/introduction-to-smart-contracts.rst:
```diff
@@ -208,7 +208,8 @@ The functions that make up the contract, and that users and contracts can call a
 The :ref:`require <assert-and-require>` function call defines conditions that reverts all changes if not met.
 In this example, ``require(msg.sender == minter);`` ensures that only the creator of the contract can call ``mint``,
-and ``require(amount < 1e60);`` ensures a maximum amount of tokens. This ensures that there are no overflow errors in the future.
+and ``require(amount < 1e60);`` ensures balance isn't overflowed by any feasible number of calls (``uint``
+maximum value exceeds 1e77.)
```
What about something like "ensures a maximum amount of tokens: The function can be called very often, but due to gas costs, there is still a practical limit."?

waded replied:
I like that as a turn away from overflow concerns and toward how the economy works - there's no room for a primer on how to handle error/overflow in smart contracts here. I would still keep the tip about how this value compares to maximum uint value though - I showed someone else this and he commented that 1e60 seemed to be a huge number. Not that huge, though! Let me revise.

waded committed:
Incorporate review feedback (gas vs. overflow)

As chriseth alludes to in review, it's not so much that overflow is prevented, but that the gas cost (which should get someone googling what that is if they missed it), the cost to mint repeatedly, prohibits many calls. Because I asked else what they thought and got feedback they didn't realize how huge the uint max value is, compared to 1e60 ... "can it overflow in just one call?" - I think it's also reasonable to keep the context parenthetical about uint max here.

waded committed:
This is more clear as an "and"

The parenthesis allowed this to become an "or" for some readers.

waded committed:
Fix typo in code quoting. Remove extra semicolon

waded committed:
Restore semicolon for consistency across file

chriseth commented on docs/introduction-to-smart-contracts.rst:
```diff
@@ -208,7 +208,8 @@ The functions that make up the contract, and that users and contracts can call a
 In this example, ``require(msg.sender == minter);`` ensures that only the creator of the contract can call ``mint``,
-and ``require(amount < 1e60);`` ensures a maximum amount of tokens. This ensures that there are no overflow errors in the future.
+and ``require(amount < 1e60);`` ensures a maximum amount of tokens: The function can be called very often,
+but due to gas costs, there is still a practical limit, and ``uint`` maximum value exceeds 1e77.
```
I don't understand - how is `1e77` relevant here?

chriseth replied:
Also, I think we should keep the comment about overflow, since this is the reason the restriction is there in the first place.

waded replied:
Perhaps you've changed your mind about the overflow mention, since your prior suggestion "ensures a maximum amount of tokens: The function can be called very often, but due to gas costs, there is still a practical limit" removed it. Maybe in combination with my addition we could say "ensures a maximum amount of coins: The function can be called very often, but due to gas costs, there is still a practical limit, with overflow of ``uint amount`` only occurring in excess of 1e77 coins minted."

(I replaced the word "token" you used with "coins" here, as the document calls the unit of amount "coins" only.)

Providing the maximum value of uint is, as I mentioned in PR description, critical to understanding potential for overflow if it's to be discussed here. Overflow prevention in most programming doesn't focus on one operand, as was done here, and is maybe reasonable in crypto programming when it's not elsewhere. But although this doc mentions uints are "unsigned integer of 256 bits" earlier, astronomically large numbers are hard, and I think not many will go "oh right, that's WAY bigger than 1e60" unless stated right here, and it's easy enough to state.

chriseth replied:
By not mentioning it, I did not mean to suggest it to be deleted, sorry.

You suggestion sounds good, but maybe we could also add the exact number of calls required for overflow?

Also it's probably better to split the first sentence into two, since what follows only concerns the second require.

hrkrshnn replied:
Wouldn't it be easier if we removed the line `require(amount < 1e60)`? The safemath would prevent `balances[receiver] += amount;` from overflowing anyway. At least as an introduction, it's better to avoid this rather tricky idea that with the check `require(amount < 1e60)`, it would take more than `10**17` calls to `mint` to get an overflow (which wouldn't happen anyway because of safemath). And this is impractical because of gas reasons.

waded replied:
@hrkrshnn Totally agree. It was my first thought opening this PR, but commit history showed a couple chances to have removed that code, so not being a contributor here I took an "improve before remove" approach - but as @chriseth and I have walked through revisions, I think there's no explanation simple enough for an intro. Were the line of code and comment about overflow not there at all I'd have breezed right through this part of the sample. @chriseth if you're OK with this, I'll just remove the LOC and the related bit of commentary.

chriseth replied:
Yep, sounds good! I would still prefer some documentation about potential overflows, though, since people should be aware.

waded replied:
@chriseth I've updated the PR to remove instead. I've not read most of the rest of the docs, very new to Solidarity - looking around briefly, I see pretty clear warning about integers being restricted to ranges in docs/types.md, what happens on overflow. I also see the SMTChecker article which makes me think there's a way to shake out overflows before they happen if I'm taking this contract to production - that were I doing this production, might make me want to introduce a line or lines like `require(amount < 1e60)`.

Perhaps linking to types.md closer to initial discussion of uint in this intro would be sufficient?

waded committed:
Remove check and overflow commentary

Although overflow checking is certainly a Solidarity coding concern, as discussed in review it's simpler to not get into partial/simplistic discussion of it here with an LOC that may/may-not solve it.

hrkrshnn commented on docs/introduction-to-smart-contracts.rst:
```diff
@@ -207,9 +206,7 @@ The functions that make up the contract, and that users and contracts can call a
-In this example, ``require(msg.sender == minter);`` ensures that only the creator of the contract can call ``mint``,
-and ``require(amount < 1e60);`` ensures a maximum amount of tokens: The function can be called very often,
-but due to gas costs, there is still a practical limit, and ``uint`` maximum value exceeds 1e77.
+In this example, ``require(msg.sender == minter);`` ensures that only the creator of the contract can call ``mint``.
```
```suggestion
In this example, ``require(msg.sender == minter);`` ensures that only the creator of the contract can call ``mint``. Note that because of the default :ref:`Checked arithmetic <_unchecked>`, the transaction would revert if the expression ``balances[receiver] += amount;`` overflows, i.e., when ``balances[receiver] + amount`` in arbitrary precision arithmetic is larger than the maximum value of ``uint`` (``2**256 - 1``).
```

hrkrshnn replied:
@waded Already committed this change. I would say this PR can now be merged if you can squash all the commits. Could you please squash? Also thanks for the PR.

chriseth replied:
Shouldn't the second part go to the description of `send`?

hrkrshnn committed:
Update docs/introduction-to-smart-contracts.rst

hrkrshnn commented on docs/introduction-to-smart-contracts.rst:
```diff
@@ -207,8 +206,7 @@ The functions that make up the contract, and that users and contracts can call a
 The :ref:`require <assert-and-require>` function call defines conditions that reverts all changes if not met.
-In this example, ``require(msg.sender == minter);`` ensures that only the creator of the contract can call ``mint``,
-and ``require(amount < 1e60);`` ensures a maximum amount of tokens. This ensures that there are no overflow errors in the future.
+In this example, ``require(msg.sender == minter);`` ensures that only the creator of the contract can call ``mint``. Note that because of the default :ref:`Checked arithmetic <_unchecked>`, the transaction would revert if the expression ``balances[receiver] += amount;`` overflows, i.e., when ``balances[receiver] + amount`` in arbitrary precision arithmetic is larger than the maximum value of ``ui ...
```
```suggestion
In this example, ``require(msg.sender == minter);`` ensures that only the creator of the contract can call ``mint``. Note that because of the default :ref:`Checked arithmetic <unchecked>`, the transaction would revert if the expression ``balances[receiver] += amount;`` overflows, i.e., when ``balances[receiver] + amount`` in arbitrary precision arithmetic is larger than the maximum value of ``uint`` (``2**256 - 1``).
```

hrkrshnn committed:
Update docs/introduction-to-smart-contracts.rst

waded commented:
No meaning to the close-and-reopen - hit the CLOSE button in GitHub mobile app by mistake. Thanks for ignoring!

hrkrshnn commented on docs/introduction-to-smart-contracts.rst:
```diff
@@ -207,8 +206,7 @@ The functions that make up the contract, and that users and contracts can call a
 The :ref:`require <assert-and-require>` function call defines conditions that reverts all changes if not met.
-In this example, ``require(msg.sender == minter);`` ensures that only the creator of the contract can call ``mint``,
-and ``require(amount < 1e60);`` ensures a maximum amount of tokens. This ensures that there are no overflow errors in the future.
+In this example, ``require(msg.sender == minter);`` ensures that only the creator of the contract can call ``mint``. Note that because of the default :ref:`Checked arithmetic <unchecked>`, the transaction would revert if the expression ``balances[receiver] += amount;`` overflows, i.e., when ``balances[receiver] + amount`` in arbitrary precision arithmetic is larger than the maximum value of ``uin ...
```
```suggestion
In this example, ``require(msg.sender == minter);`` ensures that only the creator of the contract can call ``mint``. Note that because of the default :ref:`Checked arithmetic <unchecked>`, the transaction would revert if the expression ``balances[receiver] += amount;`` overflows, i.e., when ``balances[receiver] + amount`` in arbitrary precision arithmetic is larger than the maximum value of ``uint`` (``2**256 - 1``). This is also true for the statement ``balances[receiver] += amount;`` in the function ``send``.
```

hrkrshnn replied:
Added this extra line.

hrkrshnn committed:
Update docs/introduction-to-smart-contracts.rst

hrkrshnn commented:
@waded could you please squash all the 11 commits to 1. This can be merged afterwards.

chriseth commented on docs/introduction-to-smart-contracts.rst:
```diff
@@ -207,8 +206,7 @@ The functions that make up the contract, and that users and contracts can call a
 The :ref:`require <assert-and-require>` function call defines conditions that reverts all changes if not met.
-In this example, ``require(msg.sender == minter);`` ensures that only the creator of the contract can call ``mint``,
-and ``require(amount < 1e60);`` ensures a maximum amount of tokens. This ensures that there are no overflow errors in the future.
+In this example, ``require(msg.sender == minter);`` ensures that only the creator of the contract can call ``mint``. Note that because of the default :ref:`Checked arithmetic <unchecked>`, the transaction would revert if the expression ``balances[receiver] += amount;`` overflows, i.e., when ``balances[receiver] + amount`` in arbitrary precision arithmetic is larger than the maximum value of ``uin ...
```
Please add a paragraph in the rendered document and add line breaks in the source code.

Maybe we should also introduce this sentence with something like `In general, the creator can mint as many tokens as they like, but at some point, this will lead to a phenomenon called "overflow":`

chriseth replied:
Note that this is the "introduction to smart contracts" sections - we assume readers have not ever seen a smart contract before in their life, or not even a program in general.

hrkrshnn commented:
Replaced by https://github.com/ethereum/solidity/pull/11314
